### PR TITLE
OrientDB studio

### PR DESCRIPTION
**Pull request by jdillon**
https://issues.sonatype.org/browse/NEXUS-7857

Adds JMX plugin ( https://issues.sonatype.org/browse/NEXUS-7358 ), debatably useful since the JMX integration exposes non-portable mbeans.

Adds configuration of http protocol and network listener

Hacked in studio.zip into bundle, not happy with it... but it works.

Edit `etc/nexus.properties`:

```
nexus.orient.httpListenerEnabled=true
nexus.orient.dynamicPlugins=true
```

Start server and browse to:

* http://localhost:2480/studio/index.html

Auth is admin/admin, its annoyingly popping up browser auth until you tell it, seems to have a bunch of quirks, but may still be useful to help development.

https://www.evernote.com/l/ABK_vzYlRZBCKrfKx4Vvdzxlst44n0qPsVc

https://www.evernote.com/l/ABKnwOXujj5AfJLlL7BagY0gSSWkXAFMkZc

The 'sonatype-work/nexus/orient/plugins' directory is not otherwise controllable, we may want to get them to make this configurable, right now its always `${orient.home}/plugins`.

### CODE REVIEW

**Comment by jdillon** [comment]
Seems it spits out some nonsense too:

```
2015-07-12 14:33:44,057-0700 ERROR [OrientDB HTTP Connection /127.0.0.1:2480<-/127.0.0.1:65190] *SYSTEM com.orientechnologies.orient.server.network.protocol.http.ONetworkProtocolHttpDb - Internal server error:
com.orientechnologies.orient.core.exception.OQueryParsingException: Error on parsing query at position #8: Error on parsing command at position #0: Error on parsing query
Query:  _studio WHERE user.name = 'admin' AND type = 'GraphConfig'
-------------^
--> com.orientechnologies.orient.core.exception.OCommandExecutionException: Class '_STUDIO' was not found in current database
2015-07-12 14:33:44,072-0700 WARN  [OrientDB HTTP Connection /127.0.0.1:2480<-/127.0.0.1:65190] *SYSTEM com.orientechnologies.orient.server.network.protocol.http.ONetworkProtocolHttpDb - ->127.0.0.1: Command not found: GET.sqlProfiler/security
```

**Comment by kellyrob99** [comment]
+1 should be much handier than crafting queries on the command line

**Comment by bradbeck** [comment]
Could the `studio.zip` be added via a bundle fragment hosted by an OrientDB bundle (e.g. `com.orientechnologies.orientdb-server`)?

**Comment by jdillon** [comment]
@bradbeck that file has to live in that exact location for orient to find it and load it, anything more than overlaying onto the dist means we have to write code to otherwise dynamically extract it or register the plugin.

This artifact also only exists in the orientdb distribution and not easily accessible from any other public repo AFAICT.

**Comment by lvca** [comment]
By the way, not OPair is serializable, so JMX results are portable.

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java [hunk 1]
```diff
@@ -58,16 +65,24 @@
...
+  private boolean dynamicPlugins;
 
   private OServer server;
 
   @Inject
   public DatabaseServerImpl(final ApplicationDirectories applicationDirectories,
-                            final @Named("${nexus.orient.listenerEnabled:-false}") boolean listenerEnabled)
+                            final @Named("${nexus.orient.binaryListenerEnabled:-false}") boolean binaryListenerEnabled,
+                            final @Named("${nexus.orient.httpListenerEnabled:-false}") boolean httpListenerEnabled,
+                            final @Named("${nexus.orient.dynamicPlugins:-false}") boolean dynamicPlugins)
```
Comment (sonatype-zion) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java [hunk 2]
```diff
@@ -76,10 +91,17 @@ public DatabaseServerImpl(final ApplicationDirectories applicationDirectories,
...
+  public boolean isBinaryListenerEnabled() {
+    return binaryListenerEnabled;
+  }
+
+  @ManagedAttribute
+  public boolean isHttpListenerEnabled() {
+    return httpListenerEnabled;
   }
 
+  // FIXME: May need to revisit embedded configuration strategy, this is quite nasty
```
Comment (sonatype-zion) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: MAJOR Take the required action to fix the issue indicated by this comment. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java [hunk 3]
```diff
@@ -114,35 +136,72 @@ private OServerConfiguration createConfiguration() {
...
+      listener.protocol = "binary";
+      listener.socket = "default";
+      config.network.listeners.add(listener);
+      log.info("Binary listener enabled: {}:[{}]", listener.ipAddress, listener.portRange);
+    }
+
+    // HACK: Optionally enable the http listener
+    if (httpListenerEnabled) {
+      OServerNetworkListenerConfiguration listener = new OServerNetworkListenerConfiguration();
+      listener.ipAddress = "0.0.0.0";
```
Comment (sonatype-zion) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: MAJOR Make this IP "0.0.0.0" address configurable. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1313)

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java [hunk 4]
```diff
@@ -58,16 +65,24 @@
...
+
+  private boolean dynamicPlugins;
 
   private OServer server;
 
   @Inject
   public DatabaseServerImpl(final ApplicationDirectories applicationDirectories,
-                            final @Named("${nexus.orient.listenerEnabled:-false}") boolean listenerEnabled)
+                            final @Named("${nexus.orient.binaryListenerEnabled:-false}") boolean binaryListenerEnabled,
+                            final @Named("${nexus.orient.httpListenerEnabled:-false}") boolean httpListenerEnabled,
```
Comment (sonatype-zion) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

**Comment by sonatype-zion** [comment]
SonarQube analysis reported 4 issues:
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java (hunk 2 above)
Comment (jdillon) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: ignoring for now

File: components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java (hunk 3 above)
Comment (jdillon) [inline components/nexus-core/src/main/java/org/sonatype/nexus/internal/orient/DatabaseServerImpl.java]: ignoring for now

**Comment by sonatype-zion** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by mcculls** [comment]
+1 note: will need some minor reworking wrt. the upcoming working directory change

**Comment by jdillon** [comment]
manually merged, will adjust resources in assembly after nexus-work change.

**Pull request closed without merging**